{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First read --------> \n",
      "Name:  C0HVYACXX120402:7:1207:5722:57044  [ 76 ]\n",
      "Reference name:  21\n",
      "Cigar string:  76M\n",
      "Aligment position:  0 : 76\n",
      "Sequnce :  TTTTCAAACA\n",
      "Reference:  21\n",
      "---- Flags: \n",
      "Unmapped:  False\n",
      "Reverse complemented:  False\n",
      "Secondary aligment:  False\n",
      "Supplementary aligment:  False\n",
      "Not passing quality controls:  False\n",
      "\n",
      "\n",
      "Random read  2511865  -------->\n",
      "Name:  C0HVYACXX120402:1:2202:6125:126119  [ 76 ]\n",
      "Reference name:  22\n",
      "Cigar string:  76M\n",
      "Aligment position:  0 : 76\n",
      "Sequnce :  CCCAAGTTGC\n",
      "Reference:  22\n",
      "---- Flags: \n",
      "Unmapped:  False\n",
      "Reverse complemented:  True\n",
      "Secondary aligment:  False\n",
      "Supplementary aligment:  False\n",
      "Not passing quality controls:  False\n",
      "\n",
      "\n",
      "Random read  1785021  -------->\n",
      "Name:  C0HVYACXX120402:6:1302:7895:49889  [ 76 ]\n",
      "Reference name:  22\n",
      "Cigar string:  76M\n",
      "Aligment position:  0 : 76\n",
      "Sequnce :  CAGCTTCTCA\n",
      "Reference:  22\n",
      "---- Flags: \n",
      "Unmapped:  False\n",
      "Reverse complemented:  True\n",
      "Secondary aligment:  False\n",
      "Supplementary aligment:  False\n",
      "Not passing quality controls:  False\n",
      "\n",
      "\n",
      "Random read  2689853  -------->\n",
      "Name:  C0HVYACXX120402:6:2206:12890:29805  [ 76 ]\n",
      "Reference name:  22\n",
      "Cigar string:  76M\n",
      "Aligment position:  0 : 76\n",
      "Sequnce :  AGGGCTGGTC\n",
      "Reference:  22\n",
      "---- Flags: \n",
      "Unmapped:  False\n",
      "Reverse complemented:  True\n",
      "Secondary aligment:  False\n",
      "Supplementary aligment:  False\n",
      "Not passing quality controls:  False\n",
      "\n",
      "\n",
      "\n",
      "Calculations:\n",
      "\n",
      "Unmapped reads:  17765\n",
      "Total reads:  2921629\n",
      "Reads with zero quality:  126628\n",
      "Average quality for all :  55.9137915868\n",
      "Average without zero quality reads :  58.4469755109\n"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "def check_flag(value, mask):\n",
    "    return bool(value & mask)\n",
    "\n",
    "def examine_flags(flagValue):\n",
    "    #print custom subset of all flags\n",
    "    print('Unmapped: ' , check_flag(flagValue, 0x04))\n",
    "    print('Reverse complemented: ' , check_flag(flagValue, 0x10))\n",
    "    print('Secondary aligment: ' , check_flag(flagValue, 0x100))\n",
    "    print('Supplementary aligment: ' , check_flag(flagValue, 0x800))\n",
    "    print('Not passing quality controls: ' , check_flag(flagValue, 0x200))\n",
    "\n",
    "#Print fields in AlignedSegment \n",
    "def examine_read(read):\n",
    "    #print custom read fields\n",
    "    print('Name: ', read.query_name , ' [' , read.query_length, ']')\n",
    "    print('Reference name: ' , read.reference_name)\n",
    "    print('Cigar string: ' , read.cigarstring)\n",
    "    print('Aligment position: ' , read.query_alignment_start , ':' , read.query_alignment_end)\n",
    "    print('Sequnce : ', read.query_sequence[:10]) #print only first 10 nucleotides\n",
    "    print('Reference: ' , read.reference_name)\n",
    "    print('---- Flags: ')\n",
    "    examine_flags(read.flag)\n",
    "\n",
    "\n",
    "#load bam file\n",
    "bamFile = pysam.AlignmentFile('/sbgenomics/project-files/merged-tumor.bam', 'rb') #mode rb to read bam\n",
    "\n",
    "#create list of reads from bam file\n",
    "reads = []\n",
    "unmapped = 0\n",
    "for read in bamFile : \n",
    "    reads.append(read)\n",
    "    if( read.is_unmapped ):\n",
    "        unmapped += 1\n",
    "        \n",
    "#get first read, AlignedSegment type\n",
    "firstRead = reads[0]\n",
    "print('First read --------> ')\n",
    "examine_read(firstRead)\n",
    "\n",
    "#print info about 3 more random reads\n",
    "for i in range(3):\n",
    "    randomValue = random.randint(0, len(reads) - 1 )\n",
    "    read = reads[randomValue]\n",
    "    print('\\n\\nRandom read ', randomValue, ' -------->')\n",
    "    examine_read(read)\n",
    "    \n",
    "# General calucalations:\n",
    "print('\\n\\n\\nCalculations:\\n')\n",
    "print('Unmapped reads: ', unmapped)\n",
    "\n",
    "print('Total reads: ', len(reads))\n",
    "\n",
    "#Performances aren't our top priority (all of bellow we can calculate in intial for loop)\n",
    "\n",
    "# drop reads with non zero quality \n",
    "nonZeroQ = [read for read in reads if read.mapping_quality > 0] # or filter(lambda read: read.mapping_quality > 0, reads)\n",
    "print('Reads with zero quality: ', len(reads) - len(nonZeroQ))\n",
    "\n",
    "totalMean = np.mean([read.mapping_quality for read in reads])\n",
    "print('Average quality for all : ', totalMean)\n",
    "\n",
    "nonZeroMean = np.mean([read.mapping_quality for read in nonZeroQ])\n",
    "print('Average without zero quality reads : ', nonZeroMean)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
